packages/react-native/React/Base/RCTRootView.m

```diff
@@ -438,6 +439,63 @@ - (void)viewDidChangeEffectiveAppearance
 #endif // macOS]
 
 
+#pragma mark - Key window blur/focus
+
+#if TARGET_OS_OSX // [macOS
+- (void)viewDidMoveToWindow {
+  [super viewDidMoveToWindow];
+
+  NSWindow *window = [self window];
+  if (window == nil) {
+    return;
+  }
+
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(containingWindowDidBecomeKey)
+                                               name:NSWindowDidBecomeKeyNotification
+                                             object:window];
+
+  [[NSNotificationCenter defaultCenter] addObserver:self
+                                           selector:@selector(containingWindowDidResignKey)
+                                               name:NSWindowDidResignKeyNotification
+                                             object:window];
+}
+
++ (NSNumber *)reactTagClosestToView:(RCTPlatformView *)view {
+  // The first responder may not necessarily have a React tag. For example, if we're focused on
+  // a multiline <TextInput>, the RN component whose blur/focus events we need to touch is an
+  // RCTMultilineTextInputView, but the first responder will be the underlying RCTUITextView.
+  for (RCTPlatformView *testView = view; testView != nil; testView = [testView superview]) {
```

Comment:
yikes... what's the perf impact on this? should we... cache react tag?

Reply:
In practice, not that much. This only gets called when we change the key window, and most React Native components are fairly self-contained, so we'd only climb up a few levels of the view hierarchy.